cmd/iam.go

```diff
@@ -374,20 +374,73 @@ func (sys *IAMSys) watch(ctx context.Context) {
 	}
 
 	// Fall back to loading all items periodically
-	ticker := time.NewTicker(sys.iamRefreshInterval)
-	defer ticker.Stop()
+
+	// allocate dynamic timeout once before the loop
+	iamLockTimeout := newDynamicTimeout(5*time.Second, 3*time.Second)
 	for {
-		select {
-		case <-ticker.C:
-			if err := sys.Load(ctx); err != nil {
-				logger.LogIf(ctx, fmt.Errorf("Failure in periodic refresh for IAM: %v", err))
+		// we are calling this function directly (so we can use defer
+		// calls and cleanup easily).
+		func() {
+			nCtx, cancel := context.WithCancel(ctx)
+			defer cancel()
+			r := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+			// If we get this lock we will hold it and send IAM
+			// updates via peer calls.
+			watcherLock := objAPI.NewNSLock(minioMetaBucket, minioConfigPrefix+"/iam-refresh.lock")
+			lkctx, err := watcherLock.GetLock(nCtx, iamLockTimeout)
+			if err != nil {
+				// Sleep for a random interval at most 5 seconds.
+				time.Sleep(time.Duration(r.Float64() * float64(5*time.Second)))
+				return
 			}
-		case <-ctx.Done():
-			return
-		}
+			// Got the lock!
+			defer watcherLock.Unlock(lkctx.Cancel)
+
+			// Periodically refresh IAM and notify all peers.
+			ticker := time.NewTicker(sys.iamRefreshInterval)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ticker.C:
```

Comment:
This ticker will tick immediately, aren't we loading IAM in the startup ?

Reply:
After entering the select block, it will tick only after another refresh interval time - why do you say it will tick immediately?

Tested here: https://go.dev/play/p/EdeWfdz_s4D

Reply:
I also tested here - https://go.dev/play/p/QWASOFfODzE

If you see, it will tick immedialtey, not after 10 seconds

Reply:
Huh!

I see this output from your script:

```
Start time: 2009-11-10 23:00:00 +0000 UTC m=+0.000000001
Current time:  2009-11-10 23:00:10 +0000 UTC m=+10.000000001
Current time:  2009-11-10 23:00:20 +0000 UTC m=+20.000000001
Done!

Program exited.
```

First tick is at 23:00:10 which is 10 seconds later.